File: final-project/pitcher/android/app/src/main/kotlin/dev/rohanjsh/pitcher/cast/session/CastSessionManager.kt
package dev.rohanjsh.pitcher.cast.session

import android.util.Log
import dev.rohanjsh.pitcher.cast.CastConnectionState
import dev.rohanjsh.pitcher.cast.CastDevice
import dev.rohanjsh.pitcher.cast.CastPlaybackState
import dev.rohanjsh.pitcher.cast.MediaInfo
import dev.rohanjsh.pitcher.cast.provider.CastProviderContract
import dev.rohanjsh.pitcher.cast.provider.CastProviderObserver


class CastSessionManager : CastProviderObserver {

    fun interface StateObserver {
        fun onStateChanged(state: SessionSnapshot)
    }

    fun interface DevicesObserver {
        fun onDevicesChanged(devices: List<CastDevice>)
    }

    private val providers = mutableMapOf<String, CastProviderContract>()

    private var activeProvider: CastProviderContract? = null
    private var currentState = SessionSnapshot()

    private var stateObserver: StateObserver? = null
    private var devicesObserver: DevicesObserver? = null

    fun registerProvider(provider: CastProviderContract) {
        log("Registering provider: ${provider.identifier}")
        providers[provider.identifier] = provider
        provider.setObserver(this)
    }

    fun startDiscovery() {
        log("Starting discovery on ${providers.size} providers")
        providers.values.forEach { it.startDiscovery() }
    }

    fun stopDiscovery() {
        log("Stopping discovery")
        providers.values.forEach { it.stopDiscovery() }
    }

    fun getDiscoveredDevices(): List<CastDevice> {
        return providers.values.flatMap { it.getDiscoveredDevices() }
    }

    fun connect(deviceId: String) {
        val provider = resolveProviderForDevice(deviceId)
        if (provider == null) {
            log("No provider found for device: $deviceId")
            updateState(currentState.copy(
                connectionState = CastConnectionState.DISCONNECTED,
                playbackState = CastPlaybackState.ERROR,
                errorMessage = "Device not found: $deviceId"
            ))
            return
        }
        
        activeProvider?.takeIf { it.identifier != provider.identifier }?.disconnect()

        log("Connecting via ${provider.identifier} to device: $deviceId")
        activeProvider = provider
        provider.connect(deviceId)
    }

    fun disconnect() {
        log("Disconnect requested")
        activeProvider?.disconnect()
        activeProvider = null
        updateState(SessionSnapshot())
    }

    fun loadMedia(mediaInfo: MediaInfo, autoplay: Boolean, positionMs: Long) {
        guardActive("loadMedia") { it.loadMedia(mediaInfo, autoplay, positionMs) }
    }

    fun play() = guardActive("play") { it.play() }

    fun pause() = guardActive("pause") { it.pause() }

    fun seek(positionMs: Long) = guardActive("seek") { it.seek(positionMs) }

    fun stop() = guardActive("stop") { it.stop() }

    fun setVolume(volume: Double) = guardActive("setVolume") { it.setVolume(volume) }

    fun setMuted(muted: Boolean) = guardActive("setMuted") { it.setMuted(muted) }

    fun setStateObserver(observer: StateObserver?) {
        stateObserver = observer
    }

    fun setDevicesObserver(observer: DevicesObserver?) {
        devicesObserver = observer
    }

    fun dispose() {
        log("Disposing session manager")
        providers.values.forEach { it.dispose() }
        providers.clear()
        activeProvider = null
        stateObserver = null
        devicesObserver = null
    }

    override fun onProviderStateChanged(provider: CastProviderContract, state: SessionSnapshot) {
        if (activeProvider != null && activeProvider?.identifier != provider.identifier) {
            log("Ignoring state from inactive provider: ${provider.identifier}")
            return
        }

        if (state.connectionState == CastConnectionState.CONNECTED && activeProvider == null) {
            activeProvider = provider
        }

        if (state.connectionState == CastConnectionState.DISCONNECTED) {
            if (activeProvider?.identifier == provider.identifier) {
                activeProvider = null
            }
        }

        val newState = if (state.connectionState == CastConnectionState.DISCONNECTED) {
            state.copy(activeProviderId = null)
        } else {
            state
        }
        updateState(newState)
    }

    override fun onProviderDevicesChanged(provider: CastProviderContract, devices: List<CastDevice>) {
        log("Devices changed from ${provider.identifier}: ${devices.size} devices")
        devicesObserver?.onDevicesChanged(getDiscoveredDevices())
    }

    private fun resolveProviderForDevice(deviceId: String): CastProviderContract? {
        return providers.values.find { provider ->
            provider.getDiscoveredDevices().any { it.id == deviceId }
        }
    }

    private inline fun guardActive(operation: String, action: (CastProviderContract) -> Unit) {
        val provider = activeProvider
        if (provider == null) {
            log("No active provider for $operation")
            return
        }
        action(provider)
    }

    private fun updateState(newState: SessionSnapshot) {
        if (currentState != newState) {
            log("State changed: $newState")
            currentState = newState
            stateObserver?.onStateChanged(newState)
        }
    }

    private fun log(message: String) = Log.d(TAG, message)

    companion object {
        private const val TAG = "CastSessionManager"
    }
}

data class SessionSnapshot(
    val connectionState: CastConnectionState = CastConnectionState.DISCONNECTED,
    val playbackState: CastPlaybackState = CastPlaybackState.IDLE,
    val connectedDevice: CastDevice? = null,
    val activeProviderId: String? = null,
    val positionMs: Long = 0,
    val durationMs: Long = 0,
    val errorMessage: String? = null
)

File: final-project/pitcher/ios/Runner/Cast/Session/CastSessionManager.swift
import Foundation
import os.log

@MainActor
final class CastSessionManager: CastProviderObserver {

    typealias StateObserver = (SessionSnapshot) -> Void
    typealias DevicesObserver = ([CastDevice]) -> Void

    private var providers: [String: CastProviderContract] = [:]
    private var activeProvider: CastProviderContract?
    private var currentState = SessionSnapshot()

    private var stateObserver: StateObserver?
    private var devicesObserver: DevicesObserver?

    func registerProvider(_ provider: CastProviderContract) {
        log("Registering provider: \(provider.identifier)")
        providers[provider.identifier] = provider
        provider.setObserver(self)
    }

    // MARK: - Discovery

    func startDiscovery() {
        log("Starting discovery on \(providers.count) providers")
        providers.values.forEach { $0.startDiscovery() }
    }

    func stopDiscovery() {
        log("Stopping discovery")
        providers.values.forEach { $0.stopDiscovery() }
    }

    func getDiscoveredDevices() -> [CastDevice] {
        providers.values.flatMap { $0.getDiscoveredDevices() }
    }

    // MARK: - Connection

    func connect(deviceId: String) {
        guard let provider = resolveProviderForDevice(deviceId) else {
            log("No provider found for device: \(deviceId)")
            updateState(currentState.copy(
                connectionState: .disconnected,
                playbackState: .error,
                errorMessage: "Device not found: \(deviceId)"
            ))
            return
        }

        if let active = activeProvider, active.identifier != provider.identifier {
            active.disconnect()
        }

        log("Connecting via \(provider.identifier) to device: \(deviceId)")
        activeProvider = provider
        provider.connect(deviceId: deviceId)
    }

    func disconnect() {
        log("Disconnect requested")
        activeProvider?.disconnect()
        activeProvider = nil
        updateState(SessionSnapshot())
    }

    // MARK: - Playback

    func loadMedia(_ mediaInfo: MediaInfo, autoplay: Bool, positionMs: Int64) {
        guardActive("loadMedia") { $0.loadMedia(mediaInfo, autoplay: autoplay, positionMs: positionMs) }
    }

    func play() {
        guardActive("play") { $0.play() }
    }

    func pause() {
        guardActive("pause") { $0.pause() }
    }

    func seek(positionMs: Int64) {
        guardActive("seek") { $0.seek(positionMs: positionMs) }
    }

    func stop() {
        guardActive("stop") { $0.stop() }
    }

    func setVolume(_ volume: Double) {
        guardActive("setVolume") { $0.setVolume(volume) }
    }

    func setMuted(_ muted: Bool) {
        guardActive("setMuted") { $0.setMuted(muted) }
    }

    // MARK: - Observers

    func setStateObserver(_ observer: StateObserver?) {
        stateObserver = observer
    }

    func setDevicesObserver(_ observer: DevicesObserver?) {
        devicesObserver = observer
    }

    func dispose() {
        log("Disposing session manager")
        providers.values.forEach { $0.dispose() }
        providers.removeAll()
        activeProvider = nil
        stateObserver = nil
        devicesObserver = nil
    }

    // MARK: - CastProviderObserver

    func onProviderStateChanged(_ provider: CastProviderContract, state: SessionSnapshot) {
        if activeProvider != nil && activeProvider?.identifier != provider.identifier {
            log("Ignoring state from inactive provider: \(provider.identifier)")
            return
        }

        if state.connectionState == .connected && activeProvider == nil {
            activeProvider = provider
        }

        if state.connectionState == .disconnected {
            if activeProvider?.identifier == provider.identifier {
                activeProvider = nil
            }
        }

        let newState = state.connectionState == .disconnected
            ? state.copy(activeProviderId: nil)
            : state

        updateState(newState)
    }

    func onProviderDevicesChanged(_ provider: CastProviderContract, devices: [CastDevice]) {
        log("Devices changed from \(provider.identifier): \(devices.count) devices")
        devicesObserver?(getDiscoveredDevices())
    }

    // MARK: - Private

    private func resolveProviderForDevice(_ deviceId: String) -> CastProviderContract? {
        if deviceId.hasPrefix("airplay") {
            return providers[CastProviderIdentifiers.airPlay]
        }

        return providers.values.first { provider in
            provider.getDiscoveredDevices().contains { $0.id == deviceId }
        }
    }

    private func guardActive(_ operation: String, action: (CastProviderContract) -> Void) {
        guard let provider = activeProvider else {
            log("No active provider for \(operation)")
            return
        }
        action(provider)
    }

    private func updateState(_ newState: SessionSnapshot) {
        guard currentState != newState else { return }
        log("State changed: \(newState)")
        currentState = newState
        stateObserver?(newState)
    }

    private func log(_ message: String) {
        os_log("[CastSessionManager] %{public}@", log: .default, type: .debug, message)
    }
}
